chapter_recurrent-neural-networks/rnn-concise.ipynb

```diff
@@ -0,0 +1,808 @@
+    "        this.prepared = false;\n",
+    "    }\n",
+    "\n",
+    "    @Override\n",
+    "    public Record get(NDManager manager, long index) throws IOException {\n",
+    "        NDArray X = data.get(new NDIndex(\"{}\", index));\n",
+    "        NDArray Y = labels.get(new NDIndex(\"{}\", index));\n",
+    "        return new Record(new NDList(X), new NDList(Y));\n",
+    "    }\n",
+    "\n",
+    "    @Override\n",
+    "    protected long availableSize() {\n",
+    "        return data.getShape().get(0);\n",
+    "    }\n",
+    "\n",
+    "    @Override\n",
+    "    public void prepare(Progress progress) throws IOException, TranslateException {\n",
+    "        if (prepared) {\n",
+    "            return;\n",
+    "        }\n",
+    "\n",
+    "        Pair<List<Integer>, Vocab> corpusVocabPair = null;\n",
+    "        try {\n",
+    "            corpusVocabPair = TimeMachine.loadCorpusTimeMachine(maxTokens);\n",
+    "        } catch (Exception e) {\n",
+    "            e.printStackTrace(); // Exception can be from unknown token type during tokenize() function.\n",
+    "        }\n",
+    "        List<Integer> corpus = corpusVocabPair.getKey();\n",
+    "        this.vocab = corpusVocabPair.getValue();\n",
+    "\n",
+    "        // Start with a random offset (inclusive of `numSteps - 1`) to partition a\n",
+    "        // sequence\n",
+    "        int offset = new Random().nextInt(numSteps);\n",
+    "        int numTokens = ((int) ((corpus.size() - offset - 1) / batchSize)) * batchSize;\n",
+    "        NDArray Xs =\n",
+    "                manager.create(\n",
+    "                        corpus.subList(offset, offset + numTokens).stream()\n",
+    "                                .mapToInt(Integer::intValue)\n",
+    "                                .toArray());\n",
+    "        NDArray Ys =\n",
+    "                manager.create(\n",
+    "                        corpus.subList(offset + 1, offset + 1 + numTokens).stream()\n",
+    "                                .mapToInt(Integer::intValue)\n",
+    "                                .toArray());\n",
+    "        Xs = Xs.reshape(new Shape(batchSize, -1));\n",
+    "        Ys = Ys.reshape(new Shape(batchSize, -1));\n",
+    "        int numBatches = (int) Xs.getShape().get(1) / numSteps;\n",
+    "\n",
+    "        for (int i = 0; i < numSteps * numBatches; i += numSteps) {\n",
+    "            NDArray X = Xs.get(new NDIndex(\":, {}:{}\", i, i + numSteps));\n",
+    "            NDArray Y = Ys.get(new NDIndex(\":, {}:{}\", i, i + numSteps));\n",
+    "            // Temp variables to be able to detach NDArray which will be replaced\n",
+    "            NDArray temp = this.data;\n",
+    "            NDArray temp2 = this.data;\n",
+    "            this.data = this.data.concat(X);\n",
+    "            this.labels = this.labels.concat(Y);\n",
+    "            temp.detach();\n",
```

Comment:
You can create a NDList and do only one concat and close all the NDArray in the NDList. Also, by calling detach(), this NDArray become unmanaged and can't be freed by any NDManager. Is it what you want?

Reply:
Ahh good catch, yeah what I meant was to close it so not sure why I didn't use `close()` and used `detach()` instead. Yep, I'll go ahead and do the NDList which sounds like a better approach.